src/Raven.Server/Documents/ShardedHandlers/ShardedBatchHandler.cs

```diff
@@ -0,0 +1,355 @@
+                        var requestExecutor = ShardedContext.RequestExecutors[shardIndex];
+
+                        if (shardedBatchCommands.ContainsKey(shardIndex) == false)
+                        {
+                            shardedBatchCommands.Add(shardIndex, new SingleNodeShardedBatchCommand(this, requestExecutor.ContextPool));
+                        }
+
+                        var shardedBatchCommand = shardedBatchCommands[shardIndex];
+                        shardedBatchCommand.AddCommand(command);
+                    }
+
+                    var tasks = shardedBatchCommands.Select(x => ShardedContext.RequestExecutors[x.Key].ExecuteAsync(x.Value, x.Value.Context));
+                    await Task.WhenAll(tasks);
+
+                    HttpContext.Response.StatusCode = (int)HttpStatusCode.Created;
+                    using (var writer = new BlittableJsonTextWriter(context, ResponseBodyStream()))
+                    {
+                        var reply = new DynamicJsonArray();
+                        foreach (var command in shardedBatchCommands.Values)
+                        {
+                            var result = command.Result;
+                            result.TryGet(nameof(BatchCommandResult.Results), out BlittableJsonReaderArray partialResult);
+                            foreach (var o in partialResult.Items)
+                            {
+                                reply.Add(o);
+                            }
+                        }
+                        context.Write(writer, new DynamicJsonValue {[nameof(BatchCommandResult.Results)] = reply});
+                    }
+                }
+            }
+        }
+    }
+
+    public class SingleNodeShardedBatchCommand : ShardedCommand
+    {
+        private readonly JsonOperationContext _context;
+        private readonly List<Stream> _commands = new List<Stream>();
+        private List<Stream> _attachmentStreams;
+        private HashSet<Stream> _uniqueAttachmentStreams;
+        private readonly TransactionMode _mode = TransactionMode.SingleNode;
+        private readonly IDisposable _returnCtx;
+
+        public JsonOperationContext Context => _context;
+
+        public SingleNodeShardedBatchCommand(ShardedRequestHandler handler, JsonContextPool pool) :
+            base(handler, ShardedCommands.Headers.None)
+        {
+            _returnCtx = pool.AllocateOperationContext(out _context);
+        }
+
+        public void AddCommand(SingleShardedCommand command)
+        {
+            _commands.Add(command.BufferedCommand);
+
+            if (command.AttachmentStream != null)
+            {
+                var stream = command.AttachmentStream;
+                if (_attachmentStreams == null)
+                {
+                    _attachmentStreams = new List<Stream>();
+                    _uniqueAttachmentStreams = new HashSet<Stream>();
+                }
+
+                PutAttachmentCommandHelper.ValidateStream(stream);
```

Comment:
Why is this required?

Reply:
It doesn't cost us anything and I think we actually need it to be able to put it on the network stream.
```
public static void ValidateStream(Stream stream)
        {
            if (stream.CanRead == false)
                ThrowNotReadableStream();
            if (stream.CanSeek == false)
                ThrowNotSeekableStream();
            if (stream.Position != 0)
                ThrowPositionNotZero(stream.Position);
        }
```